// @genql/cli/dist/runtime/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.everything = exports.GenqlError = exports.createFetcher = exports.linkTypeMap = exports.generateGraphqlOperation = exports.createClient = void 0;
var createClient_1 = require("./createClient");
Object.defineProperty(exports, "createClient", { enumerable: true, get: function () { return createClient_1.createClient; } });
var generateGraphqlOperation_1 = require("./generateGraphqlOperation");
Object.defineProperty(exports, "generateGraphqlOperation", { enumerable: true, get: function () { return generateGraphqlOperation_1.generateGraphqlOperation; } });
var linkTypeMap_1 = require("./linkTypeMap");
Object.defineProperty(exports, "linkTypeMap", { enumerable: true, get: function () { return linkTypeMap_1.linkTypeMap; } });
// export { Observable } from 'zen-observable-ts'
var fetcher_1 = require("./fetcher");
Object.defineProperty(exports, "createFetcher", { enumerable: true, get: function () { return fetcher_1.createFetcher; } });
var error_1 = require("./error");
Object.defineProperty(exports, "GenqlError", { enumerable: true, get: function () { return error_1.GenqlError; } });
exports.everything = {
    __scalar: true,
};
