// @genql/cli/dist/render/common/comment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.argumentComment = exports.fieldComment = exports.typeComment = exports.comment = void 0;
const comment = (comment) => {
    const lines = [];
    if (comment.deprecated) {
        lines.push(`@deprecated ${comment.deprecated.replace(/\s/g, ' ')}`);
    }
    if (comment.text) {
        lines.push(...comment.text.split('\n'));
    }
    return lines.length > 0
        ? lines.length === 1
            ? `\n/** ${lines[0]} */\n`
            : `\n/**\n${lines.map(l => ` * ${l}`).join('\n')}\n */\n`
        : '';
};
exports.comment = comment;
const typeComment = (type) => (0, exports.comment)({
    text: type.description,
});
exports.typeComment = typeComment;
const fieldComment = (field) => (0, exports.comment)({
    deprecated: field.deprecationReason,
    text: field.description,
});
exports.fieldComment = fieldComment;
const argumentComment = (arg) => (0, exports.comment)({
    text: arg.description,
});
exports.argumentComment = argumentComment;
